language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,410 @@
+            .filter_map(|(r, a)| match r.into_inner() {
+                MethodResponse::GetAccountStateWithProof(v) => {
+                    let state = self.extract_account_state(&v);
+                    Some(AccountAddressWithState { address: *a, state })
+                }
+                _ => None,
+            })
+            .collect();
+
+        Ok(account_states)
+    }
+
+    fn get_addresses_read_by_txs(
+        &self,
+        txs: &[Transaction],
+        replay_version: u64,
+    ) -> Result<Vec<AccountAddress>> {
+        // TODO: use dependency analysis to fetch the modules relevant
+        // to the transactions rather than all the modules at address 0x1
+        let modules = self
+            .debugger
+            .get_diem_framework_modules_at_version(replay_version, true)?;
+
+        let mut keys = vec![];
+        let rw_analysis = ReadWriteSetAnalysis::new(read_write_set::analyze(&modules)?);
+
+        let live_state_view = DebuggerStateView::new(self.validator.as_ref(), replay_version);
+        let live_storage = RemoteStorage::new(&live_state_view);
+
+        for tx in txs.iter() {
+            let new_keys = self.get_keys_read_for_tx(&live_storage, tx, &rw_analysis)?;
+            keys.extend(new_keys);
+        }
+        Ok(self.get_unique_addresses_from_keys(&keys))
+    }
+
+    fn populate_storage_for_transactions(
+        &self,
+        txs: &[Transaction],
+        version: Option<u64>,
+    ) -> Result<()> {
+        let replay_version = match version {
+            Some(0) => 1,
+            Some(v) => v,
+            None => self.debugger.get_latest_version()?,
+        };
+
+        let addresses_read = self.get_addresses_read_by_txs(txs, replay_version)?;
+        let accounts_states = self.fetch_accounts_storage(&addresses_read)?;
+        for account in accounts_states {
+            self.debugger
+                .save_account_state(account.address, &account.state)?;
+        }
+        Ok(())
+    }
+
+    fn _execute_transactions_at<V: VMExecutor>(
+        &self,
+        state: &OnDiskStateViewReader,
+        txs: Vec<Transaction>,
+        version: Option<u64>,
+    ) -> Result<Vec<TransactionOutput>> {
+        self.populate_storage_for_transactions(&txs, version)?;
+        V::execute_block(txs, state).map_err(|err| format_err!("Unexpected VM Error: {:?}", err))
+
+        // let live_state_view =
+        //     DebuggerStateView::new(self.validator.as_ref(), self.debugger.get_latest_version()?);
+        // V::execute_block(txs, &live_state_view)
+        //     .map_err(|err| format_err!("Unexpected VM Error: {:?}", err))
+    }
+
+    fn cleanup_dirs(&self) -> Result<()> {
+        let dirs_to_clean = vec![&self.config.storage_dir, &self.config.build_dir];
+        dirs_to_clean.into_iter().fold(Ok(()), |acc, dir| {
+            let res = fs::remove_dir_all(dir);
+            acc.and(res.map_err(|e| format_err!("could not remove {:?}: {}", dir, e)))
+        })
```

Comment:
It's a bit hard to tell exactly what the semantics are here. Would this work instead? 
```suggestion
       for dir in dirs_to_clean.into_iter() {
           fs::remove_dir_all(dir).map_err(|e| format_err!("could not remove {:?}: {}", dir, e))
        }
```

Reply:
I think this would change the semantics a bit because here we don't want an early-return but I changed this to a more digestible imperative style